RealmSwift/Map.swift

```diff
@@ -0,0 +1,610 @@
+ 
+ Map only supports String as a key.
+ 
+ Unlike Swift's native collections, `Map`s is a reference types, and are only immutable if the Realm that manages them
+ is opened as read-only.
+ 
+ A Map can be filtered and sorted with the same predicates as `Results<Value>`.
+ 
+ Properties of `Map` type defined on `Object` subclasses must be declared as `let` and cannot be `dynamic`.
+*/
+public final class Map<Key, Value>: RLMSwiftCollectionBase where Key: _MapKey, Value: RealmCollectionValue {
+
+    // MARK: Properties
+
+    /// The Realm which manages the map, or `nil` if the map is unmanaged.
+    public var realm: Realm? {
+        return _rlmCollection.realm.map { Realm($0) }
+    }
+
+    /// Indicates if the map can no longer be accessed.
+    public var isInvalidated: Bool { return _rlmCollection.isInvalidated }
+
+    /// Returns all of the keys in this map.
+    public var keys: [Key] {
+        return rlmDictionary.allKeys.map(dynamicBridgeCast)
+    }
+
+    /// Returns all of the values in this map.
+    public var values: [Value] {
+        return rlmDictionary.allValues.map(dynamicBridgeCast)
+    }
+
+    // MARK: Initializers
+
+    /// Creates a `Map` that holds Realm model objects of type `Value`.
+    public override init() {
+        super.init()
+    }
+
+    internal init(objc rlmDictionary: RLMDictionary<AnyObject, AnyObject>) {
+        super.init(collection: rlmDictionary)
+    }
+
+    // MARK: Count
+
+    /// Returns the number of key-value pairs in this map.
+    @objc public var count: Int { return Int(_rlmCollection.count) }
+
+    // MARK: Mutation
+
+    /**
+     Updates the value stored in the map for the given key, or adds a new key-value pair if the key does not exist.
+
+     - parameter value: a value's key path predicate.
+     - parameter forKey: The direction to sort in.
+     */
+    public func updateValue(_ value: Value, forKey key: Key) {
+        rlmDictionary[objcKey(from: key)] = dynamicBridgeCast(fromSwift: value) as AnyObject
+    }
+
+    /**
+     Removes the given key and its associated object.
+     */
+    public func removeObject(for key: Key) {
+        rlmDictionary.removeObject(for: objcKey(from: key))
+    }
+
+    /**
+     Removes all objects from the map. The objects are not removed from the Realm that manages them.
+
+     - warning: This method may only be called during a write transaction.
+     */
+    public func removeAll() {
+        rlmDictionary.removeAllObjects()
+    }
+
+    /// :nodoc:
+    public subscript(key: Key) -> Value? {
+        get {
+            if rlmDictionary.type == .object {
+                let obj = rlmDictionary[objcKey(from: key)]
+                // A Map can keep the key of an object that has been deleted by the Realm.
+                // If the object is deleted it will be stored as NSNull.null so we want to
+                // return that as `nil`.
+                if obj is NSNull {
+                    return nil
+                }
```

Comment:
This is incorrectly conflating "value for this key is null" and "dictionary does not have an entry for this key".  There shouldn't be any need to special-case object here at all; the only thing different about object from other types is that object-valued Maps must be optional, and the handling of option values of any type should be the same.